### ISSUE TITLE
improve memory usage for large files

### ISSUE THREAD

**Issue by jtmoon79** [jtmoon79/super-speedy-syslog-searcher#182]
## Summary

This is a "meta issue" to link to specific issues around a chronic issue.

### Current behavior

Currently, some files are entirely read into memory. When dealing with large files and/or many files then the process can use too much memory and possibly crash due to Out Of Memory errors.

Reading from a large amount of archived files was the original motivating use-case for this project, i.e. many tars of logs from related systems. It's too bad it's not more fully supported.

Specifically, these circumstances:
* `.xz` files must be [entirely read into memory during `Blockreader::new`](https://github.com/jtmoon79/super-speedy-syslog-searcher/blob/0.6.66/src/readers/blockreader.rs#L1037-L1117)<br/>
  see #12
* `.tar` file contained file must be [entirely read into memory](https://github.com/jtmoon79/super-speedy-syslog-searcher/blob/0.6.66/src/readers/blockreader.rs#L2307-L2356) during the first call to `BlockReader::read_block_FileTar`<br/>
  see #13<br/>
  In other words, the file contained within the `.tar` file is entirely read into memory, e.g. the entire file `syslog` from `logs.tar` but not the entire file `logs.tar`
* `*.evtx` files with "out of order" events must be [entirely read into memory](https://github.com/jtmoon79/super-speedy-syslog-searcher/blob/0.6.68/src/readers/evtxreader.rs#L182-L203)<br/>
  relates to #86

### Suggested behavior

Have some O(1) ceiling for memory usage for all cases.

Relates to #14.

**Comment by jtmoon79** [comment]
# The current implementation

### `SyslineReader::find_sysline_at_datetime_filter` binary search

The `SyslogProcessor::find_sysline_between_datetime_filters` calls `SyslineReader::find_sysline_at_datetime_filter`
which does a binary search over the entire file during the "find datetime" stage (`Stage2FindDt`). This stage occurs before any log messages
are printed to the console (during stage `Stage3StreamSyslines`).
The intention here is that when a user
passes a `--dt-after` value then the first qualifying value can be found quickly by skipping the reading and
processing of unnecessary data (any syslines before the `--dt-after` datetime). This was one of the
original insights that motivated writing a custom `BlockReader` and it's attendant classes; very large log
files read over slow communication links could be search very quickly for log messages between some narrow band of time.

### underlying `BlockReader` and drops

The underlying `BlockReader` does not know which stage is currently in progress (as it should not; stages are only a concept to the `SyslogProcessor`). So the `BlockReader` does not know when it should drop any held blocks. Nor could the `BlockReader` decide to drop it's own blocks because the user of the `BlockReader`, the `LineReader` holds `BlockP` points to the `Block`s held by the `BlockReader`. In other words, even if the `BlockReader` decided to be clever and drop some of the `Block`s it is storing, there would still be pointers to those `Block`s (an `Arc<Vec<u8>>`) held by the `LineReader` instance. And the `BlockReader` should not try to get into signaling the `LineReader` to drop those `BlockP` pointers (a `BlockReader` should know nothing about a `LineReader`).

Signaling the `BlockReader` to drop data is currently initiated by the driver function in `bin.rs`, `exec_syslogprocessor`.
The calls to `SyslogProcessor::drop_data_try` is done during the log printing stage `Stage3StreamSyslines`.

To review: during the binary search done early on during stage `Stage2FindDt` (again, only needed if the user passes `--dt-after`) there is one call to `SyslineReader::find_sysline_at_datetime_filter`. There are no attempts to tell the `BlockReader` to drop old data.
And during this stage, it is possible for an entire `.xz` file or an entire `.gz` file to be held in memory if, for example, the user passes `--dt-after` datetime value that is after all log messages in the decompressed log data.

# Idea: BlockReader only holds _N_ blocks at a time

So the next idea would be to modify the `BlockReader` so that it only ever holds some _N_ amount of `Block`s

### Problem

But that's difficult because remember the controlling `LineReader` hold it's own copies of pointers to those `Block`s as well (`BlockP` instances) so the underyling `Vec<u8>` wouldn't drop as there are outstanding pointers to it. Additionally, it would be a bad design to try to have the `BlockReader` callback to the `LineReader` (now the `BlockReader` needs to understand what a `LineReader` is and that get's too messy).

# Idea: BlockReader proactively drops blocks when handling `xz` files

Another idea is to modify the `BlockReader::read_block_FileXz` such that it procatively drops blocks that are not needed. Function `read_block_FileXz` does a linear search, storing `Block`s as they are decompressed, and only passes back one `BlockP` when it is done. So the `BlockReader` could quietly drop "old" `Block`s of data as new `Block`s are decompressed.
For example, given a call to `read_block_FileXz(9)` to read the tenth `Block` of data, that requires decompressing all prior `Block`s of data (as xz can only perform a "stream" of compression/decompression). The function could be modified such that as `Block` two is read then `Block` one is dropped, and so on, up to the tenth `Block` (so nine prior decompressed `Block`s would have been dropped).

### Problem

But then the problem is another call from the overlying `SyslineProcessor` within binary search function `find_sysline_at_datetime_filter` may later call `read_block_FileXz(5)`. If `Block`s one through six were droped previously, then the entire file must read and decompressed again (reading the sixth `Block` requires reading the prior five `Block`s). So the initial binary search with _O(log(n))_ time now has _O(n*log(n))_ time.
